src/PrestaShopBundle/Security/Admin/UserProvider.php

```diff
@@ -0,0 +1,151 @@
+{
+    /**
+     * @var Context
+     */
+    private $legacyContext;
+
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $cache;
+
+    /**
+     * @var CommandBusInterface
+     */
+    private $queryBus;
+
+    /**
+     * @param LegacyContext $context
+     * @param CacheItemPoolInterface $cache
+     * @param CommandBusInterface $queryBus
+     */
+    public function __construct(
+        LegacyContext $context,
+        CacheItemPoolInterface $cache,
+        CommandBusInterface $queryBus
+    ) {
+        $this->legacyContext = $context->getContext();
+        $this->cache = $cache;
+        $this->queryBus = $queryBus;
+    }
+
+    /**
+     * Fetch the Employee entity that matches the given username.
+     * Cache system doesn't supports "@" character, so we rely on a sha1 expression.
+     *
+     * @param string $username
+     *
+     * @return Employee
+     *
+     * @throws InvalidArgumentException
+     * @throws UsernameNotFoundException
+     */
+    public function loadUserByUsername($username)
+    {
+        $cacheKey = sha1($username);
+        $cachedEmployee = $this->cache->getItem("app.employees_${cacheKey}");
+
+        if ($cachedEmployee->isHit()) {
+            return $cachedEmployee->get();
+        }
+
+        try {
+            /** @var EmployeeForAuthentication $employeeForAuthentication */
+            $employeeForAuthentication = $this->queryBus->handle(
+                GetEmployeeForAuthentication::fromEmail($username)
+            );
+            $employee = new Employee(
+                new LegacyEmployee($employeeForAuthentication->getEmployeeId()->getValue())
+            );
+            $employee->setRoles($employeeForAuthentication->getRoles());
+
+            $cachedEmployee->set($employee);
+            $this->cache->save($cachedEmployee);
+
+            return $cachedEmployee->get();
+        } catch (AuthenticatingEmployeeNotFoundException $e) {
+            throw new UsernameNotFoundException(
+                sprintf('Username "%s" does not exist.', $username),
+                0,
+                $e
+            );
+        }
+    }
+
+    /**
+     * Reload an Employee and returns a fresh instance.
+     *
+     * @param UserInterface $employee
+     *
+     * @return Employee
+     */
+    public function refreshUser(UserInterface $employee)
+    {
+        if (!$employee instanceof Employee) {
+            throw new UnsupportedUserException(
+                sprintf('Instances of "%s" are not supported.', get_class($employee))
+            );
+        }
+
+        return $this->loadUserByUsername($employee->getUsername());
+    }
+
+    /**
+     * Tests if the given class supports the security layer.
+     * Here, only Employee class is allowed to be used to authenticate.
+     *
+     * @param string $class
+     *
+     * @return bool
+     */
+    public function supportsClass($class)
+    {
+        return $class === Employee::class;
```

Comment:
This would be better to use `instanceof` no?

Reply:
again, this class is coupled with legacy `Employee`, mostly because of BC.

Reply:
`$class` is a string class name so I would need to instantiate an object to do `instanceof`, which I think is worse in this case :thinking:

Reply:
Apparently `instanceof` does not do that, but [is_subclass_of](https://www.php.net/manual/en/function.is-subclass-of.php) does the job. Same for `is_a`

Reply:
are these suggestions any better though? I feel like this class name comparison is common practice in Symfony context, as it's even used in the Symfony docs.